File: flash_cards.py


#!/usr/bin/env python3
# flash_cards.py
import argparse
import json
import csv
import os
import sys
import random
from datetime import datetime
from colorama import init, Fore, Style

init(autoreset=True)

DATA_FILE = "cards.json"

class Card:
    def __init__(self, question, answer, card_set="Общее", correct=0, wrong=0):
        self.question = question
        self.answer = answer
        self.set = card_set
        self.correct = correct
        self.wrong = wrong

    def to_dict(self):
        return {
            "question": self.question,
            "answer": self.answer,
            "set": self.set,
            "correct": self.correct,
            "wrong": self.wrong
        }

    @classmethod
    def from_dict(cls, data):
        return cls(data["question"], data["answer"], data.get("set", "Общее"),
                   data.get("correct", 0), data.get("wrong", 0))

    def get_id(self, all_cards):
        # Для удобства используем индекс в списке
        pass

class FlashCards:
    def __init__(self):
        self.cards = []
        self.load()

    def load(self):
        if not os.path.exists(DATA_FILE):
            return
        try:
            with open(DATA_FILE, 'r') as f:
                data = json.load(f)
                self.cards = [Card.from_dict(item) for item in data]
        except:
            self.cards = []

    def save(self):
        with open(DATA_FILE, 'w') as f:
            json.dump([c.to_dict() for c in self.cards], f, indent=2)

    def add_card(self, question, answer, card_set="Общее"):
        self.cards.append(Card(question, answer, card_set))
        self.save()
        print(Fore.GREEN + f"Карточка добавлена (ID: {len(self.cards)-1})")

    def remove_card(self, idx):
        if idx < 0 or idx >= len(self.cards):
            print(Fore.RED + "Неверный ID.")
            return
        removed = self.cards.pop(idx)
        self.save()
        print(Fore.YELLOW + f"Карточка '{removed.question}' удалена.")

    def list_cards(self):
        if not self.cards:
            print(Fore.YELLOW + "Нет карточек.")
            return
        print(Fore.CYAN + "📚 Список карточек:")
        for i, c in enumerate(self.cards):
            print(f"  {i}. {c.question} | {c.answer} | набор: {c.set} | правильные: {c.correct}, ошибки: {c.wrong}")

    def train(self, card_set=None):
        cards = self.cards
        if card_set:
            cards = [c for c in cards if c.set == card_set]
            if not cards:
                print(Fore.RED + f"Набор '{card_set}' не найден.")
                return
        if not cards:
            print(Fore.YELLOW + "Нет карточек для тренировки.")
            return
        random.shuffle(cards)
        correct_total = 0
        wrong_total = 0
        print(Fore.GREEN + f"Начинаем тренировку (набор: {card_set if card_set else 'все'})")
        print("Введите ответ, или 'q' для выхода, или 's' для пропуска.")
        for card in cards:
            print(Fore.CYAN + f"Вопрос: {card.question}")
            user_input = input(Fore.YELLOW + "Ваш ответ: " + Style.RESET_ALL).strip()
            if user_input.lower() == 'q':
                print(Fore.MAGENTA + "Тренировка прервана.")
                break
            if user_input.lower() == 's':
                print(Fore.BLUE + f"Правильный ответ: {card.answer}")
                continue
            if user_input.lower() == card.answer.lower():
                print(Fore.GREEN + "✅ Правильно!")
                card.correct += 1
                correct_total += 1
            else:
                print(Fore.RED + f"❌ Неправильно. Правильный ответ: {card.answer}")
                card.wrong += 1
                wrong_total += 1
            self.save()
        total = correct_total + wrong_total
        if total > 0:
            print(Fore.MAGENTA + f"Тренировка завершена. Правильных: {correct_total}, ошибок: {wrong_total}, точность: {correct_total/total*100:.1f}%")

    def stats(self):
        if not self.cards:
            print(Fore.YELLOW + "Нет данных.")
            return
        total_correct = sum(c.correct for c in self.cards)
        total_wrong = sum(c.wrong for c in self.cards)
        total = total_correct + total_wrong
        print(Fore.CYAN + "📊 Статистика:")
        print(f"  Всего карточек: {len(self.cards)}")
        print(f"  Правильных ответов: {total_correct}")
        print(f"  Неправильных ответов: {total_wrong}")
        if total > 0:
            print(f"  Точность: {total_correct/total*100:.1f}%")
        # Карточки с низкой точностью
        weak = [c for c in self.cards if c.correct + c.wrong > 0 and (c.correct / (c.correct + c.wrong) < 0.5)]
        if weak:
            print(Fore.YELLOW + "Слабые карточки (точность < 50%):")
            for c in weak:
                acc = c.correct / (c.correct + c.wrong) * 100
                print(f"  {c.question} | точность: {acc:.1f}%")

    def export_json(self, filename):
        with open(filename, 'w') as f:
            json.dump([c.to_dict() for c in self.cards], f, indent=2)
        print(Fore.GREEN + f"Экспортировано в {filename} (JSON)")

    def import_json(self, filename):
        try:
            with open(filename, 'r') as f:
                data = json.load(f)
                for item in data:
                    self.cards.append(Card.from_dict(item))
            self.save()
            print(Fore.GREEN + f"Импортировано {len(data)} карточек из {filename}")
        except Exception as e:
            print(Fore.RED + f"Ошибка импорта: {e}")

    def export_csv(self, filename):
        with open(filename, 'w', newline='') as f:
            writer = csv.writer(f)
            writer.writerow(["question", "answer", "set", "correct", "wrong"])
            for c in self.cards:
                writer.writerow([c.question, c.answer, c.set, c.correct, c.wrong])
        print(Fore.GREEN + f"Экспортировано в {filename} (CSV)")

def main():
    parser = argparse.ArgumentParser(description="Тренер памяти (карточки)")
    parser.add_argument("--add", help="Добавить карточку (вопрос)")
    parser.add_argument("--answer", help="Ответ на карточку")
    parser.add_argument("--set", default="Общее", help="Набор (категория)")
    parser.add_argument("--remove", type=int, help="Удалить карточку по ID")
    parser.add_argument("--list", action="store_true", help="Показать все карточки")
    parser.add_argument("--train", help="Начать тренировку (указать набор, или без аргумента - все)")
    parser.add_argument("--stats", action="store_true", help="Показать статистику")
    parser.add_argument("--export-json", help="Экспорт в JSON")
    parser.add_argument("--import-json", help="Импорт из JSON")
    parser.add_argument("--export-csv", help="Экспорт в CSV")
    args = parser.parse_args()

    fc = FlashCards()

    if args.add:
        if not args.answer:
            print(Fore.RED + "Для добавления карточки требуется --answer")
            sys.exit(1)
        fc.add_card(args.add, args.answer, args.set)
    elif args.remove is not None:
        fc.remove_card(args.remove)
    elif args.list:
        fc.list_cards()
    elif args.train is not None:
        fc.train(args.train if args.train != "" else None)
    elif args.stats:
        fc.stats()
    elif args.export_json:
        fc.export_json(args.export_json)
    elif args.import_json:
        fc.import_json(args.import_json)
    elif args.export_csv:
        fc.export_csv(args.export_csv)
    else:
        parser.print_help()

if __name__ == "__main__":
    main()

File: flash_cards.js
#!/usr/bin/env node
// flash_cards.js
const { program } = require('commander');
const fs = require('fs');
const chalk = require('chalk');

const DATA_FILE = 'cards.json';

class Card {
    constructor(question, answer, set, correct = 0, wrong = 0) {
        this.question = question;
        this.answer = answer;
        this.set = set || 'Общее';
        this.correct = correct;
        this.wrong = wrong;
    }
}

class FlashCards {
    constructor() {
        this.cards = [];
        this.load();
    }

    load() {
        try {
            if (fs.existsSync(DATA_FILE)) {
                const data = JSON.parse(fs.readFileSync(DATA_FILE, 'utf8'));
                this.cards = data.map(item => new Card(item.question, item.answer, item.set, item.correct, item.wrong));
            }
        } catch (e) {
            this.cards = [];
        }
    }

    save() {
        fs.writeFileSync(DATA_FILE, JSON.stringify(this.cards, null, 2));
    }

    addCard(question, answer, set) {
        this.cards.push(new Card(question, answer, set));
        this.save();
        console.log(chalk.green(`Карточка добавлена (ID: ${this.cards.length - 1})`));
    }

    removeCard(idx) {
        if (idx < 0 || idx >= this.cards.length) {
            console.log(chalk.red('Неверный ID.'));
            return;
        }
        const removed = this.cards.splice(idx, 1)[0];
        this.save();
        console.log(chalk.yellow(`Карточка '${removed.question}' удалена.`));
    }

    listCards() {
        if (this.cards.length === 0) {
            console.log(chalk.yellow('Нет карточек.'));
            return;
        }
        console.log(chalk.cyan('📚 Список карточек:'));
        for (let i = 0; i < this.cards.length; i++) {
            const c = this.cards[i];
            console.log(`  ${i}. ${c.question} | ${c.answer} | набор: ${c.set} | правильные: ${c.correct}, ошибки: ${c.wrong}`);
        }
    }

    train(cardSet) {
        let cards = this.cards;
        if (cardSet) {
            cards = cards.filter(c => c.set === cardSet);
            if (cards.length === 0) {
                console.log(chalk.red(`Набор '${cardSet}' не найден.`));
                return;
            }
        }
        if (cards.length === 0) {
            console.log(chalk.yellow('Нет карточек для тренировки.'));
            return;
        }
        cards = cards.sort(() => Math.random() - 0.5);
        let correctTotal = 0, wrongTotal = 0;
        console.log(chalk.green(`Начинаем тренировку (набор: ${cardSet || 'все'})`));
        console.log('Введите ответ, или "q" для выхода, или "s" для пропуска.');
        const readline = require('readline').createInterface({
            input: process.stdin,
            output: process.stdout
        });
        const ask = (index) => {
            if (index >= cards.length) {
                readline.close();
                this._finishTraining(correctTotal, wrongTotal);
                return;
            }
            const card = cards[index];
            console.log(chalk.cyan(`Вопрос: ${card.question}`));
            readline.question(chalk.yellow('Ваш ответ: '), (answer) => {
                const userInput = answer.trim();
                if (userInput.toLowerCase() === 'q') {
                    console.log(chalk.magenta('Тренировка прервана.'));
                    readline.close();
                    return;
                }
                if (userInput.toLowerCase() === 's') {
                    console.log(chalk.blue(`Правильный ответ: ${card.answer}`));
                    ask(index + 1);
                    return;
                }
                if (userInput.toLowerCase() === card.answer.toLowerCase()) {
                    console.log(chalk.green('✅ Правильно!'));
                    card.correct++;
                    correctTotal++;
                } else {
                    console.log(chalk.red(`❌ Неправильно. Правильный ответ: ${card.answer}`));
                    card.wrong++;
                    wrongTotal++;
                }
                this.save();
                ask(index + 1);
            });
        };
        ask(0);
    }

    _finishTraining(correct, wrong) {
        const total = correct + wrong;
        if (total > 0) {
            console.log(chalk.magenta(`Тренировка завершена. Правильных: ${correct}, ошибок: ${wrong}, точность: ${(correct / total * 100).toFixed(1)}%`));
        }
    }

    stats() {
        if (this.cards.length === 0) {
            console.log(chalk.yellow('Нет данных.'));
            return;
        }
        const totalCorrect = this.cards.reduce((s, c) => s + c.correct, 0);
        const totalWrong = this.cards.reduce((s, c) => s + c.wrong, 0);
        const total = totalCorrect + totalWrong;
        console.log(chalk.cyan('📊 Статистика:'));
        console.log(`  Всего карточек: ${this.cards.length}`);
        console.log(`  Правильных ответов: ${totalCorrect}`);
        console.log(`  Неправильных ответов: ${totalWrong}`);
        if (total > 0) {
            console.log(`  Точность: ${(totalCorrect / total * 100).toFixed(1)}%`);
        }
        const weak = this.cards.filter(c => c.correct + c.wrong > 0 && (c.correct / (c.correct + c.wrong) < 0.5));
        if (weak.length > 0) {
            console.log(chalk.yellow('Слабые карточки (точность < 50%):'));
            for (const c of weak) {
                const acc = c.correct / (c.correct + c.wrong) * 100;
                console.log(`  ${c.question} | точность: ${acc.toFixed(1)}%`);
            }
        }
    }

    exportJson(filename) {
        fs.writeFileSync(filename, JSON.stringify(this.cards, null, 2));
        console.log(chalk.green(`Экспортировано в ${filename} (JSON)`));
    }

    importJson(filename) {
        try {
            const data = JSON.parse(fs.readFileSync(filename, 'utf8'));
            data.forEach(item => this.cards.push(new Card(item.question, item.answer, item.set, item.correct, item.wrong)));
            this.save();
            console.log(chalk.green(`Импортировано ${data.length} карточек из ${filename}`));
        } catch (e) {
            console.log(chalk.red(`Ошибка импорта: ${e.message}`));
        }
    }

    exportCsv(filename) {
        const header = 'question,answer,set,correct,wrong\n';
        const rows = this.cards.map(c => `${c.question},${c.answer},${c.set},${c.correct},${c.wrong}`).join('\n');
        fs.writeFileSync(filename, header + rows);
        console.log(chalk.green(`Экспортировано в ${filename} (CSV)`));
    }
}

program
    .option('--add <question>', 'Добавить карточку (вопрос)')
    .option('--answer <answer>', 'Ответ на карточку')
    .option('--set <set>', 'Набор (категория)', 'Общее')
    .option('--remove <id>', 'Удалить карточку по ID', parseInt)
    .option('--list', 'Показать все карточки')
    .option('--train [set]', 'Начать тренировку (опционально указать набор)')
    .option('--stats', 'Показать статистику')
    .option('--export-json <file>', 'Экспорт в JSON')
    .option('--import-json <file>', 'Импорт из JSON')
    .option('--export-csv <file>', 'Экспорт в CSV')
    .parse(process.argv);

const opts = program.opts();
const fc = new FlashCards();

if (opts.add) {
    if (!opts.answer) {
        console.error(chalk.red('Для добавления карточки требуется --answer'));
        process.exit(1);
    }
    fc.addCard(opts.add, opts.answer, opts.set);
} else if (opts.remove !== undefined) {
    fc.removeCard(opts.remove);
} else if (opts.list) {
    fc.listCards();
} else if (opts.train !== undefined) {
    fc.train(opts.train);
} else if (opts.stats) {
    fc.stats();
} else if (opts.exportJson) {
    fc.exportJson(opts.exportJson);
} else if (opts.importJson) {
    fc.importJson(opts.importJson);
} else if (opts.exportCsv) {
    fc.exportCsv(opts.exportCsv);
} else {
    program.help();
}
